### ISSUE TITLE
Dropdown: document.ready vs. attached

### ISSUE THREAD

**Issue by JoschaMetze** [aurelia-ui-toolkits/aurelia-materialize-bridge#200]
Hi,

if I include a md-dropdown button in my view which is visible from the beginning (no show.bind etc.) materialize is initializing the drop-down element twice. Once with the bridge code (e.g. constrain-width = false) and once at document.ready-event with the default code (empty options). This may be linked to: https://github.com/Dogfalo/materialize/blob/master/js/dropdown.js#L40

I'm not sure if it is possible to delay the initialization after the document.ready event has occured or to apply the options as html data objects so the materialize code can use the options given there.

Kind regards
Joscha

**Comment by Thanood** [comment]
Hi Joscha,

I'm not sure if I understand this correctly.

If we attached the options as data attributes wouldn't the earliest possible time be in the components `bind()` method? But that is still after `document.ready`, right?

How did you manage to get `md-dropdown` to initialize a dropdown before `md-dropdown` is started, anyway? :smile: 
Just guessing: did you add the `dropdown-button` class to the element from the start? If so, `md-dropdown` does that and then initializes the dropdown. If you don't do that, Materialize won't initialize the dropdown trigger.

Regards
Daniel

**Comment by JoschaMetze** [comment]
Hi Daniel,

not sure how I managed to do this, it might have to do with my transition to webpack and typescript :).
No, I don't use the dropdown-button class directly. <button md-button="flat: true;" md-dropdown="activates: controlBox;below-origin: true;constrain-width: false;">
It's just that the attached method of md-dropdown gets called before the document.ready event is raised in materialize.js. The expectation would be that attached is only called after the jquery event?

Kind regards
Joscha

**Comment by Thanood** [comment]
That is strange.. I wonder if this is related somehow: https://github.com/aurelia/skeleton-navigation/issues/557

Indeed, `document.ready` should happen (relatively long) before any `attached()`..
Would you mind uploading a sample to reproduce this? Also, does this happen without webpack?
I would like to have a look at the repro and if I don't get any further maybe we should open an issue in one of the webpack repos.

**Comment by Thanood** [comment]
I've tried this in a ts-webpack skeleton (no Materialize):

`index.html`:

```html
<script>
    document.addEventListener('DOMContentLoaded', function() { console.log('document.ready'); }, false);
</script>
```

`custom-attr.js`:

```javascript
import {customAttribute} from 'aurelia-framework';

@customAttribute('custom-attr')
export class CustomAttr {
    attached() {
        console.log('custom-attr.attached');
    }
}
```

`app.html`:

```html
<div custom-attr class="page-host">
  <router-view></router-view>
</div>
```

I'm getting this console output:

```
document.ready
DEBUG [aurelia] Loading plugin aurelia-templating-binding.
DEBUG [aurelia] Configured plugin aurelia-templating-binding.
DEBUG [aurelia] Loading plugin aurelia-templating-resources.
DEBUG [aurelia] Configured plugin aurelia-templating-resources.
DEBUG [aurelia] Loading plugin aurelia-history-browser.
DEBUG [aurelia] Configured plugin aurelia-history-browser.
DEBUG [aurelia] Loading plugin aurelia-templating-router.
DEBUG [aurelia] Configured plugin aurelia-templating-router.
DEBUG [aurelia] Loading plugin aurelia-event-aggregator.
DEBUG [aurelia] Configured plugin aurelia-event-aggregator.
INFO [aurelia] Aurelia Started
DEBUG [templating] importing resources for app.html ["bootstrap/dist/css/bootstrap.css", "nav-bar.html", "custom-attr"]
DEBUG [templating] importing resources for nav-bar.html []
DEBUG [templating] importing resources for welcome.html ["custom-attr"]
custom-attr.attached
```

So `document.ready` is logged even before aurelia's core plugins take control.
Now if I didn't listen for the wrong event (I didn't have a look at how jQuery determines `ready`) the document should be ready when `index.html` is loaded. After that the Aurelia bootstrapper takes control.

I really wonder what's going on..

**Comment by Thanood** [comment]
Update:

Added `dom-ready.ts`:

```javascript
export default class Bogus { }

document.addEventListener('DOMContentLoaded', function() { console.log('document.ready import'); }, false);
```

Added to `main.ts`:

```javascript
import './dom-ready';
```

Now this happens:
```
document.ready
DEBUG [aurelia] Loading plugin aurelia-templating-binding.
DEBUG [aurelia] Configured plugin aurelia-templating-binding.
DEBUG [aurelia] Loading plugin aurelia-templating-resources.
DEBUG [aurelia] Configured plugin aurelia-templating-resources.
DEBUG [aurelia] Loading plugin aurelia-history-browser.
DEBUG [aurelia] Configured plugin aurelia-history-browser.
DEBUG [aurelia] Loading plugin aurelia-templating-router.
DEBUG [aurelia] Configured plugin aurelia-templating-router.
DEBUG [aurelia] Loading plugin aurelia-event-aggregator.
DEBUG [aurelia] Configured plugin aurelia-event-aggregator.
INFO [aurelia] Aurelia Started
DEBUG [templating] importing resources for app.html Array[3]
DEBUG [templating] importing resources for nav-bar.html Array[0]
DEBUG [templating] importing resources for welcome.html Array[0]
custom-attr.attached
document.ready import
```

**Comment by Thanood** [comment]
This is really interesting..
First: congratulations on getting this thing to work with webpack. :+1: 

This could as well be "misused" for a delayed `document.ready` event. Not the worst thing for jQuery users. But how is Materialize even initialized when `md-dropdown` is attached and materialize is not imported, yet?
Is this behavior even intended?

Attaching data-attributes would help with DropDown but this would only be a (ugly in my eyes) workaround and it wouldn't help with the majority of other components.

I don't know how to solved this, yet..

**Comment by JoschaMetze** [comment]
Materialize is initialized before md-dropdown is attached, the dropDown function is executed just as expected (you could add a console.log in the Bogus constructor to see that the import worked earlier). But the document.ready event is delayed.
I haven't tried the master version but 0.8.1. I'll have to try this without webpack, should be easy to reproduce in skeleton. Otherwise we might have to check with the webpack guys if this is an expected behavior.

Thanks for your quick analysis.

**Comment by Thanood** [comment]
I didn't try if this behaves in the same way when not using webpack.
When you do, please tell me how it turns out.. :smile: 
Will also try it tomorrow.

**Comment by JoschaMetze** [comment]
The key to get it work with webpack was to include a dummy for the materialize package. I just copied empty-package. This helps the resolve errors. Including the resources in aurelia-build-resource was easy (just included everything). Most difficult was the everchanging skeleton for typescript webpack (and the windows path issues coming with that) and some other libraries.

**Comment by JoschaMetze** [comment]
It's as feared, the jspm version works as expected (not calling the document ready import event). So it has to be a webpack issue.  Although I really can't think of any reason why an event should be retriggered after an import. 
Any ideas from your side? I might create an issue on aurelia-webpack-plugin, maybe niieani has an idea.

**Comment by Thanood** [comment]
We have a repro and a comparison that works as expected. We should create an issue in skeleton-navigation. :smile:

**Comment by Thanood** [comment]
I've just creted the issue in the `skeleton-navigation` repo (as you probably see above this comment, anyway).. :smile:

**Comment by Thanood** [comment]
@JoschaMetze This is solved when using the 1.0.0 skeleton. Can you confirm?
It's unfortunately not enough to use Aurelia 1.0.0 with an older skeleton.

**Comment by JoschaMetze** [comment]
This is solved indeed. Thanks for your support on this issue.
